src/R/Components/Impl/InteractiveWorkflow/Implementation/RSessionProviderCallback.cs

```diff
@@ -0,0 +1,29 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Common.Core.Shell;
+using Microsoft.R.Components.Extensions;
+using Microsoft.R.Host.Client;
+
+namespace Microsoft.R.Components.InteractiveWorkflow.Implementation {
+    public sealed class RSessionProviderCallback : IRSessionProviderCallback {
+        private readonly ICoreShell _coreShell;
+        private readonly Lazy<IRInteractiveWorkflow> _workflowLazy;
+
+        public async Task<IntPtr> GetApplicationWindowHandleAsync() {
+            await _coreShell.SwitchToMainThreadAsync();
+            return _coreShell.ApplicationWindowHandle;
+        }
```

Comment:
Since window handle never changes we could change it to property and then prefetch and cache in VsAppShell. Won't have to be async then.